### PR TITLE
SED-2963 implement autoscaling of agents

### PR DESCRIPTION
**Pull request by jeromecomte**

### CODE REVIEW

**Review by david-stephan — APPROVED** [review]
PR approved with minor change/clarification requested

File: step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/TokenAutoscalingExecutionPlugin.java [hunk 1]
```diff
@@ -53,39 +54,70 @@ public void initializeExecutionEngineContext(AbstractExecutionEngineContext pare
...
-        TokenForecastingContext tokenForecastingContext = getTokenForecastingContext(context);
-        Map<String, Integer> tokenForecastPerPool = tokenForecastingContext.getTokenForecastPerPool();
-        logger.debug("Token forecast estimation result: " + tokenForecastPerPool);
-
-        Set<Map<String, Interest>> criteriaWithoutMatch = tokenForecastingContext.getCriteriaWithoutMatch();
-        if (!criteriaWithoutMatch.isEmpty()) {
-            String message = "No matching pool found for selection criteria: " + criteriaWithoutMatch;
-            logger.warn(message);
-            throw new PluginCriticalException(message);
+        if (isAutoscalingEnabled()) {
```
Comment (david-stephan) [inline step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/TokenAutoscalingExecutionPlugin.java:57]: @david-stephan The 2 plugins could be split, forcasting should always be active and OS side, while AutoScaling could be  a licensed plugin on EE side.

File: step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/autoscaler/TokenAutoscalingConfiguration.java [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
 
+import io.kubernetes.client.openapi.models.V1ConfigMapList;
+
 import java.util.Map;
 
 public class TokenAutoscalingConfiguration {
 
     public Map<String, Map<String, String>> availableTokenPools;
 
+    public V1ConfigMapList availableConfigMapList;
```
Comment (david-stephan) [inline step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/autoscaler/TokenAutoscalingConfiguration.java:11]: @rubij this V1ConfigMapList class is specific to the k8s driver and should not appear in this module, please refactor and remove the dependency from the pom.xml

File: step-plans/step-plans-base-artefacts/pom.xml [hunk 3]
```diff
@@ -100,6 +100,18 @@
 			<version>${slf4j.version}</version>
 		</dependency>
 
+		<dependency>
+			<groupId>io.kubernetes</groupId>
```
Comment (david-stephan) [inline step-plans/step-plans-base-artefacts/pom.xml:104]: @rubij see comment on the TokenAutoscalingConfiguration class, should be removed.

**Review by david-stephan — CHANGES_REQUESTED** [review]

File: step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/autoscaler/TokenAutoscalingConfiguration.java (hunk 2 above)
Comment (rubij) [reply step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/autoscaler/TokenAutoscalingConfiguration.java]: Done

File: step-plans/step-plans-base-artefacts/pom.xml (hunk 3 above)
Comment (rubij) [reply step-plans/step-plans-base-artefacts/pom.xml]: Done

**Review by david-stephan — APPROVED** [review]

**Pull request merged**